### PR TITLE
Improved CLI UX + a bunch of regressions fixed

### LINKED ISSUE #717: Teleport usability issues
These things are UX issues, but they would greatly slow me down if I wasn't familiar with Teleport code.

## Handling non-existent user logins

I have two trusted clusters, "master" and "appliance". Using master as my proxy:

```
$ tsh --proxy=master --cluster=appliance ssh bob@hostname
```

If `bob` does not exist on "hostname" this will silently fail. It does not print "Bob does not exist".

## Handling Unreachable nodes

Same situation as above, try this:

```
$ tsh --proxy=master --cluster=appliance ssh badhost
```

If `badhost` is not reacheable (for example teleport ports are blocked via iptables) this command will:

* It will not print a proper error message, saying "no route to badhost"
* It will cause "master" teleport daemon to drop the tunnel connection, disconnecting "appliance" (and killing all active sessions to it!)

## Access Denied Reporting

Based on either you have a session key or not, the error message is different  (and cryptic). 
With empty `.tsh`:

```bash
$ tsh --proxy=localhost ssh baduser@localhost
ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain
```

With session keys in `.tsh`:

```bash
$ tsh --proxy=localhost ssh baduser@localhost
access denied to login "bar" when connecting to localhost:3022: failed to connect to "localhost:3022@default@9598f6db-4fc4-4336-82a1-81a827049fa6"
```

## Handling Unreachable Hosts

This is inside a local cluster:

```
~/go/src/github.com/gravitational/teleport: tsh --proxy=localhost ssh badhost

ERROR REPORT:
Original Error: *trace.ConnectionProblemError dial tcp: lookup badhost: no such host
Stack Trace:
	/Users/ekontsevoy/go/src/github.com/gravitational/teleport/lib/srv/proxy.go:260 github.com/gravitational/teleport/lib/srv.(*proxySubsys).proxyToHost
	/Users/ekontsevoy/go/src/github.com/gravitational/teleport/lib/srv/proxy.go:145 github.com/gravitational/teleport/lib/srv.(*proxySubsys).start
	/Users/ekontsevoy/go/src/github.com/gravitational/teleport/lib/srv/sshserver.go:834 github.com/gravitational/teleport/lib/srv.(*Server).handleSubsystem
	/Users/ekontsevoy/go/src/github.com/gravitational/teleport/lib/srv/sshserver.go:751 github.com/gravitational/teleport/lib/srv.(*Server).dispatch
	/Users/ekontsevoy/go/src/github.com/gravitational/teleport/lib/srv/sshserver.go:717 github.com/gravitational/teleport/lib/srv.(*Server).handleSessionRequests
	/usr/local/go/src/runtime/asm_amd64.s:2087 runtime.goexit
User Message: failed to connect to server <nil>
```

To fix: remove "stack dump" and use the normal error message.

### PR DESCRIPTION
**Pull request by kontsevoy**
This commit adds several improvements to how CLI SSH login works.
This PR closes #717 (fixes #717)

- slightly improved Makefile (but BSD compatibility is lost again)
- Validated keys are added to the SSH agent [1]
- tsh will not verify host keys _twice_ anymore
- tsh will not login into proxy _twice_ on every session anymore
- error messages for "access denied" look clean now

[1] This is huge. This means that tsh login can "feed" the keys to the built-in SSH agents of the OS and OpenSSH can fetch them from there.

QUESTION: why do we even need `tsh agent` option then? ssh-agent is installed on every Linux/OSX machine and it's _much_ nicer than ours (it forks and has CLI CRUD)

### CODE REVIEW

**Comment by leif-ohlin** [comment]
Wouldn't this also fix or rather work around #611?

**Comment by kontsevoy** [comment]
@leif-ohlin #611 is a duplicate of #579 which @russjones started working on last week. Short answer is: `tsh agent` does not need `--cluster` switch because it serves credentials for the `--proxy` it is given. Those credentials should be good to access _any trusted cluster_ (that's why they're called "trusted", clusters trust each other, i.e. your creds given to you by your proxy are good for all clusters). 

The problem is that there's a bug which means the agent does not return keys properly, @russjones is looking into it and it's scheduled to land in 2.0 in a few weeks.

File: lib/client/keyagent.go [hunk 1]
```diff
@@ -154,9 +164,45 @@ func (a *LocalKeyAgent) AddKey(host string, username string, key *Key) error {
...
+	}
 	return a.Agent.Add(*agentKey)
 }
 
-func (a *LocalKeyAgent) DeleteKey(host string, username string) error {
-	return trace.Wrap(a.keyStore.DeleteKey(host, username))
+// DeleteKey deletes the user key stored for a given proxy server
+func (a *LocalKeyAgent) DeleteKey(proxyHost string, username string) error {
+	// remove the same key from the SSH Agent:
+	if a.sshAgent != nil {
```
Comment (klizhentas) [inline lib/client/keyagent.go]: can you inverse the logic here to reduce nesting?

```go
if a.sshAgent == nil {
   // ... then remove it from the local storage:
   return trace.Wrap(a.keyStore.DeleteKey(proxyHost, username))
}

localKey, _ := a.keyStore.GetKey(proxyHost, username)
if localKey == nil {
    // not found
    return nil
}
pubKey, _, _, _, err := ssh.ParseAuthorizedKey(localKey.Cert)
if err != nil {
   // log warning and return
   return nil
}

agentKeys, _ := a.sshAgent.List()
for _, agentKey := range agentKeys {
if bytes.Contains(pubKey.Marshal(), agentKey.Blob) {
   a.sshAgent.Remove(agentKey)
   break
}

File: tool/tsh/main.go [hunk 2]
```diff
@@ -29,11 +29,11 @@ import (
 	"github.com/gravitational/teleport/lib/services"
 	"github.com/gravitational/teleport/lib/session"
 	"github.com/gravitational/teleport/lib/teleagent"
+	//"github.com/gravitational/teleport/lib/teleagent"
```
Comment (klizhentas) [inline tool/tsh/main.go]: remove left over comment

**Review by klizhentas — CHANGES_REQUESTED** [review]

File: tool/tsh/main.go (hunk 2 above)
Comment (kontsevoy) [inline tool/tsh/main.go]: done

File: lib/client/keyagent.go (hunk 1 above)
Comment (kontsevoy) [inline lib/client/keyagent.go]: done

**Pull request merged**